tracing-subscriber/src/layer.rs

```diff
@@ -407,6 +420,63 @@ pub struct Layered<L, I, S = I> {
     _s: PhantomData<fn(S)>,
 }
 
+/// A [`Subscriber`] composed of a `Subscriber` wrapped by one or more
+/// [`Layer`]s and [`Filter`]s.
+///
+/// [`Layer`]: ../layer/trait.Layer.html
+/// [`Subscriber`]: https://docs.rs/tracing-core/latest/tracing_core/trait.Subscriber.html
+/// [`Filter`]: ../layer/trait.Layer.html
+#[derive(Clone, Debug)]
+pub struct Filtered<L, S, F> {
+    layer: L,
+    filter: F,
+    _s: PhantomData<fn(S)>,
+}
+
+/// `Filter` enables the filtering of composed [`Layers`].
+///
+/// [`Layer`]: ../layer/trait.Layer.html
+pub trait Filter<L, S>
+where
+    L: Layer<S>,
+    S: Subscriber + 'static,
+{
+    /// Filter on a specific span or event's metadata.
+    fn filter(&self, metadata: &Metadata<'_>, ctx: &Context<'_, S>) -> bool;
```

Comment:
I think this also needs a `register_callsite`-esque hook (maybe call it `filter_static`? or something) so that filters can indicate that they _never_ want to see something

Reply:
So that the interest level for each layer can be cached? That makes sense. I think that will be the primary interface for `Filter`.

Reply:
> So that the interest level for each layer can be cached? That makes sense. I think that will be the primary interface for `Filter`.

so that the sum _total_ interest can be cached

Reply:
if all filters will never be interested in a trace, it needs a cached `never` interest.
if any are `sometimes` interested, it probably should always get a cached `sometimes`
if all filters are `always` interested, it needs a cached `always`

Reply:
there may be some optimizations to quick-disable something when we know a given filter is the only "sometimes", but that seems more complex. i'm mostly concerned about all `always` and all `never` getting cached `always` and `never` interests, respectively...